http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheClientTlsHalfCloseTest.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.apache;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.FileStoreTlsKeyManagersProvider;
+import software.amazon.awssdk.http.HttpExecuteRequest;
+import software.amazon.awssdk.http.HttpExecuteResponse;
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.TlsKeyManagersProvider;
+import software.amazon.awssdk.http.server.MockServer;
+
+public class ApacheClientTlsHalfCloseTest extends ClientTlsAuthTestBase {
+
+    private static TlsKeyManagersProvider tlsKeyManagersProvider;
+    private static MockServer mockServer;
+    private SdkHttpClient httpClient;
+
+    private static final int TWO_MB = 2 * 1024 * 1024;
+    private static final byte[] CONTENT = new byte[TWO_MB];
+
+    @Test
+    public void errorWhenServerHalfClosesSocketWhileStreamIsOpened() throws IOException {
+        mockServer = MockServer.createMockServer(MockServer.ServerBehavior.HALF_CLOSE);
+        mockServer.startServer(tlsKeyManagersProvider);
+
+        httpClient = ApacheHttpClient.builder()
+                                     .tlsKeyManagersProvider(tlsKeyManagersProvider)
+                                     .build();
+
+        IOException exception = assertThrows(IOException.class, () -> {
+            executeHttpRequest(httpClient);
+        });
+        assertEquals("Remote end is closed.", exception.getMessage());
+        // assertEquals("Socket input is already shutdown", exception.getCause().getMessage());
```

Comment:
@dagnir 
Why aren't we passing the inputStreamException as cause ?
```
IOException e = new IOException("Remote end is closed.",inputStreamException);
```

Reply:
I'm guessing because the cause is not important (calling getInputStream on behalf of users)? I'm fine with either way.